### PR TITLE
Hybrid reference

### LINKED ISSUE #5: Hybrid matrix format reference kernels
- [ ] requires #4 
- [ ] implement device dependent methods in `core/matrix/hybrid.cpp`
- [ ] declare all kernels required by device dependent methods in `core/base/hybrid_kernels.hpp`
- [ ] define these kernels as `NOT_COMPILED` in `core/base/device_hooks/common_kernels.inc.cpp`
- [ ] define these kernels as `NOT_IMPLEMENTED` in `cpu/matrix/hybrid_kernels.cpp` and `gpu/matrix/hybrid_kernels.cu`
- [ ] provide reference implementations of these kernels in `reference/matrix/hybrid_kernels.cpp`
- [ ] add unit tests for device-dependent methods in `reference/test/matrix/hybrid_kernels.cpp`

### PR DESCRIPTION
**Pull request by yhmtsai**
This closes #5 

Add Dense <-> Hybrid
Change the strategy_type of hybrid.
It uses the nnz_per_row in host memory to compute ell and coo limit.
In this part, is computing ell and coo limit also in GPU?

### CODE REVIEW

**Comment by tcojean** [comment]
Just a quick comment, as github shows you will have to update your code and make some changes: we changed the name `CpuExecutor` (and everything related, such as the `cpu` namespace) to `OmpExecutor` and `omp` namespace. Your code should be up to date with the develop branch and contain those changes.

**Comment by yhmtsai** [comment]
@tcojean thanks for reminding.
I have done it

File: core/matrix/dense.cpp [hunk 1]
```diff
@@ -145,6 +156,33 @@ inline void conversion_helper(Ell<ValueType, IndexType> *result,
...
+          typename OperationType>
+inline void conversion_helper(Hybrid<ValueType, IndexType> *result,
+                              MatrixType *source, const OperationType &op)
+{
+    auto exec = source->get_executor();
+    Array<size_type> row_nnz(exec, source->get_size().num_rows);
+    exec->run(TemplatedOperation<
+              ValueType>::make_calculate_nonzeros_per_row_operation(source,
+                                                                    &row_nnz));
+    size_type ell_lim = zero<size_type>(), coo_lim = zero<size_type>();
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/matrix/dense.cpp]: Fix it

**Comment by yhmtsai** [comment]
@gflegar 
I have fixed them.

I change the variable name (ell_max_nonzeros_per_row and coo_nnz).
I think they are more meaningful than ell_lim and coo_lim, but I am not sure whether they are confusing with the config of the matrix.
By the way, how about change max_nonzeros_per_row to max_nnz_per_row?

For the last question, the function copies coo_nnz to the reference executor now.

**Comment by gflegar** [comment]
LGTM.

> By the way, how about change max_nonzeros_per_row to max_nnz_per_row?

Actually, we had a similar discussion here in the group a while ago, and we decided to use `stored_elements` instead of `nonzeros` and `nnz` for the purpose you're describing. The reason is - nonzeros can be a bit confusing. For example, if for some format you have the following array of values: `{1.0, 3.2, 2.3, 0, 3.4, 4.5, 0, 0}`, what does the function that gets you the "number of nonzeros" return? 5 (the number of elements that are actually not zero), 8 (the number of stored elements), or even 6 (the number of stored elements, without the padding). I can be quite confusing for the user. Thus, for function that return information based on the allocation size, and not on the actual values in the array, we decided to use `stored_elements` instead of `nnz` or `nonzeros` - this is done in other matrix formats, and it would be great if you could rename the functions here to use the same.

**Comment by yhmtsai** [comment]
@gflegar 
I have changed max_nonzeros_per_row to num_stored_elements_per_row.
I still keep calculate_max_nnz_per_row function name because it does not count the zeros.

**Pull request merged**